docs/src/DeveloperGuide/WritingAdaptor.md

```diff
@@ -0,0 +1,264 @@
+as well as defining all settings and bind them to C++ variables. The benefit of
+this approach is that all configuration parameters for the adaptor will be
+available immediately after the adaptor is loaded by the tool.
+
+The final code to manage the configuration reads:
+
+```c++
+class HelloWorldConfig
+{
+public:
+  using String = std::string;
+
+  void setup(ConfigurationManager &config)
+  {
+    config.setSectionName("Hello world configuration",
+                          "Demonstrating how configuration works.");
+    config.addParameter(message_, "message",
+                        "Message which is printed when setting the adaptor up.");
+  }
+
+  String const& message() const
+  {
+    return message_;
+  }
+
+private:
+  String message_{"Hello world"};
+};
+```
+
+With the configuration in place, the next thing we concern ourselves with is
+loading the adaptor. This is the functionality that registers the configuration
+together with an ID and a profile setup function. In our case, the setup
+function should just print a message given a `HelloWorldConfig` instance. The
+corresponding adaptor registration reads:
+
+```c++
+extern "C" void loadAdaptor(IQirAdaptorFactory *generator)
+{
+  generator->registerAdaptorComponent<HelloWorldConfig>(
+      "adaptor.hello-world",
+      [](HelloWorldConfig const &cfg, IQirAdaptorFactory * /*generator*/, Profile & /*profile*/) {
+        std::cout << "Message: " << cfg.message() << std::endl;
+      });
+}
+```
+
+In this example, we will only concern ourselves with how to use the
+configuration and we will ignore `generator` and `profile` for now. The full
+source code to this example can be found in `AdaptorExamples/HelloWorld` and it
+can be compiled through following steps (startig from the Passes root folder):
+
+```sh
+mkdir Debug
+cd Debug
+cmake ..
+make HelloWorld
+```
+
+This will generate a `HelloWorld` dynamic library with path
+`./AdaptorExamples/libHelloWorld.(dylib|so|dll)`.
+
+## Loading the adaptor
+
+Executing `qat` and loading the `libHelloWorld` library, we see that our new
+settings are added to help page:
+
+```sh
+% ./qir/qat/Apps/qat --load ./AdaptorExamples/libHelloWorld.dylib -h
+Usage: ./qir/qat/Apps/qat [options] filename
+
+...
+
+Hello world configuration - Demonstration configuration for building a adaptor boilerplate.
+
+--message                                         Message which is printed when setting the adaptor up. Default: Hello world
+
+
+...
+```
+
+For the next part, we assume that you have a QIR located in
+`../qir/qir-tests/reduction_tests/inlining-input.ll`. To test that the setup
+function is invoked upon setting the profile up, we run
+
+```sh
+./qir/qat/Apps/qat -S --adaptor-pipeline hello-world --load ./AdaptorExamples/libHelloWorld.dylib ../qir/qir-tests/reduction_tests/inlining-input.ll
+Message: Hello world
+```
+
+We note that in order for our new component to be loaded we need to add it to
+the adaptor pipeline.
```

Comment:
How is it added to the adaptor pipeline? A link or explanation would be helpful.

Reply:
It is in the above command. I've explicitly written this out.